File: apps/web/android-src/MediaButtonPlugin.java
package com.fitech.app;

import android.content.Intent;
import android.media.AudioManager;
import android.media.session.MediaSession;
import android.media.session.PlaybackState;
import android.os.Handler;
import android.os.Looper;
import android.view.KeyEvent;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "MediaButton")
public class MediaButtonPlugin extends Plugin {

    public static MediaButtonPlugin instance = null;

    private static final long TAP_WINDOW_MS = 400L;
    private MediaSession mediaSession = null;
    private int tapCount = 0;
    private long lastTapTime = 0L;
    private Handler handler = null;
    private Runnable pendingTap = null;

    @Override
    public void load() {
        instance = this;
        handler = new Handler(Looper.getMainLooper());
    }

    private void setupMediaSession() {
        AudioManager audioManager = (AudioManager)
                getContext().getSystemService(android.content.Context.AUDIO_SERVICE);

        //noinspection deprecation
        audioManager.requestAudioFocus(null, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN);

        MediaSession session = new MediaSession(getContext(), "FiTechSession");

        session.setCallback(new MediaSession.Callback() {
            @Override
            public boolean onMediaButtonEvent(Intent mediaButtonIntent) {
                //noinspection deprecation
                KeyEvent event = mediaButtonIntent.getParcelableExtra(Intent.EXTRA_KEY_EVENT);
                if (event == null) return false;
                if (event.getAction() == KeyEvent.ACTION_DOWN) {
                    return handleTap(event.getKeyCode());
                }
                return false;
            }

            @Override
            public void onPlay() { handleTap(KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE); }

            @Override
            public void onPause() { handleTap(KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE); }
        });

        PlaybackState state = new PlaybackState.Builder()
                .setActions(PlaybackState.ACTION_PLAY_PAUSE
                        | PlaybackState.ACTION_PLAY
                        | PlaybackState.ACTION_PAUSE)
                .setState(PlaybackState.STATE_PLAYING,
                        PlaybackState.PLAYBACK_POSITION_UNKNOWN, 1f)
                .build();

        session.setPlaybackState(state);
        session.setActive(true);
        mediaSession = session;
    }

    @PluginMethod
    public void startListening(PluginCall call) {
        if (mediaSession == null) {
            try {
                setupMediaSession();
            } catch (Throwable e) {
                android.util.Log.e("MediaButtonPlugin", "setupMediaSession failed: " + e);
                call.reject("MediaSession setup failed: " + e.getMessage());
                return;
            }
        }
        call.resolve();
    }

    public boolean handleTap(int keyCode) {
        if (keyCode != KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE
                && keyCode != KeyEvent.KEYCODE_HEADSETHOOK
                && keyCode != KeyEvent.KEYCODE_MEDIA_PLAY
                && keyCode != KeyEvent.KEYCODE_MEDIA_PAUSE) {
            return false;
        }

        if (handler == null) return false;

        long now = System.currentTimeMillis();
        if (pendingTap != null) handler.removeCallbacks(pendingTap);

        if (now - lastTapTime < TAP_WINDOW_MS) {
            tapCount++;
        } else {
            tapCount = 1;
        }
        lastTapTime = now;

        pendingTap = new Runnable() {
            @Override
            public void run() {
                String eventName;
                if (tapCount >= 3) eventName = "tripleTap";
                else if (tapCount == 2) eventName = "doubleTap";
                else eventName = "singleTap";
                notifyListeners(eventName, new JSObject());
                tapCount = 0;
            }
        };
        handler.postDelayed(pendingTap, TAP_WINDOW_MS);
        return true;
    }

    @Override
    protected void handleOnDestroy() {
        if (mediaSession != null) {
            mediaSession.setActive(false);
            mediaSession.release();
            mediaSession = null;
        }
        instance = null;
        handler = null;
    }
}

File: apps/web/android-src/MediaButtonPlugin.kt
package com.fitech.app

import android.content.Intent
import android.media.AudioManager
import android.media.session.MediaSession
import android.media.session.PlaybackState
import android.os.Handler
import android.os.Looper
import android.view.KeyEvent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "MediaButton")
class MediaButtonPlugin : Plugin() {

    companion object {
        private const val TAP_WINDOW_MS = 400L
        var instance: MediaButtonPlugin? = null
    }

    private var mediaSession: MediaSession? = null
    private var tapCount = 0
    private var lastTapTime = 0L
    private val handler = Handler(Looper.getMainLooper())
    private var pendingTap: Runnable? = null

    override fun load() {
        instance = this
        try {
            setupMediaSession()
        } catch (e: Exception) {
            android.util.Log.e("MediaButtonPlugin", "setupMediaSession failed", e)
        }
    }

    private fun setupMediaSession() {
        val audioManager = context.getSystemService(android.content.Context.AUDIO_SERVICE) as AudioManager

        // Request audio focus so the system routes media button events to us
        @Suppress("DEPRECATION")
        audioManager.requestAudioFocus(
            { },
            AudioManager.STREAM_MUSIC,
            AudioManager.AUDIOFOCUS_GAIN
        )

        val session = MediaSession(context, "FiTechSession")

        session.setCallback(object : MediaSession.Callback() {
            override fun onMediaButtonEvent(mediaButtonIntent: Intent): Boolean {
                val event = mediaButtonIntent.getParcelableExtra<KeyEvent>(Intent.EXTRA_KEY_EVENT)
                    ?: return false
                if (event.action == KeyEvent.ACTION_DOWN) {
                    return handleTap(event.keyCode)
                }
                return false
            }

            // Called by some earbuds for play/pause
            override fun onPlay() { handleTap(KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE) }
            override fun onPause() { handleTap(KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE) }
        })

        // Must set an active PlaybackState so Android routes media buttons to this session
        val state = PlaybackState.Builder()
            .setActions(
                PlaybackState.ACTION_PLAY_PAUSE or
                PlaybackState.ACTION_PLAY or
                PlaybackState.ACTION_PAUSE
            )
            .setState(PlaybackState.STATE_PLAYING, PlaybackState.PLAYBACK_POSITION_UNKNOWN, 1f)
            .build()

        session.setPlaybackState(state)
        session.isActive = true
        mediaSession = session
    }

    @PluginMethod
    fun startListening(call: PluginCall) {
        if (mediaSession == null) {
            try {
                setupMediaSession()
            } catch (e: Exception) {
                android.util.Log.e("MediaButtonPlugin", "setupMediaSession failed", e)
                call.reject("MediaSession setup failed: ${e.message}")
                return
            }
        }
        call.resolve()
    }

    // Also called from MainActivity.dispatchKeyEvent as a backup for wired earbuds
    fun handleTap(keyCode: Int): Boolean {
        if (keyCode != KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE &&
            keyCode != KeyEvent.KEYCODE_HEADSETHOOK &&
            keyCode != KeyEvent.KEYCODE_MEDIA_PLAY &&
            keyCode != KeyEvent.KEYCODE_MEDIA_PAUSE) {
            return false
        }

        val now = System.currentTimeMillis()
        pendingTap?.let { handler.removeCallbacks(it) }

        if (now - lastTapTime < TAP_WINDOW_MS) {
            tapCount++
        } else {
            tapCount = 1
        }
        lastTapTime = now

        pendingTap = Runnable {
            val eventName = when {
                tapCount >= 3 -> "tripleTap"
                tapCount == 2 -> "doubleTap"
                else          -> "singleTap"
            }
            notifyListeners(eventName, JSObject())
            tapCount = 0
        }
        handler.postDelayed(pendingTap!!, TAP_WINDOW_MS)
        return true
    }

    override fun handleOnDestroy() {
        mediaSession?.isActive = false
        mediaSession?.release()
        mediaSession = null
        instance = null
    }
}
